File: apps/api/app/schemas/console.py
"""Request and response models for the partner console and the MoSJE dashboard.

The queue models are where the privacy posture becomes visible. A Channel Partner
processing a loan legitimately needs to know who the applicant is — but SETU never held
the full Aadhaar to begin with, so the console shows a name, a masked ID and the last
four digits of a phone number, and the real KYC happens at the branch with the document
in hand. That is not a limitation we worked around; it is the design.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class LoginRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    email: str = Field(max_length=255, examples=["partner@setu.gov.in"])
    password: str = Field(max_length=128)


class UserOut(BaseModel):
    id: str
    email: str
    display_name: str
    role: str
    partner_id: str | None = None
    partner_name: str | None = None


class LoginResponse(BaseModel):
    access_token: str
    token_type: Literal["bearer"] = "bearer"
    expires_in_minutes: int
    user: UserOut


class ApplicantSummary(BaseModel):
    """Everything the console may show about a citizen, and nothing more."""

    display_name: str | None = None
    # Last four digits only. The full number was never stored (CLAUDE.md rule 4).
    gov_id_type: str | None = None
    gov_id_last4: str | None = None
    phone_last4: str | None = None
    district: str | None = None
    state: str | None = None
    preferred_language: str = "en"


class ReadinessOut(BaseModel):
    """How close this application is to being processable, and what is missing."""

    score: float = Field(ge=0, le=1)
    uploaded: int
    required: int
    missing: list[str]
    warnings: int
    # True where a stored document could not have its ID masked automatically.
    unmasked_documents: int = 0


class QueueItem(BaseModel):
    reference_no: str
    status: str
    scheme_code: str
    scheme_name: str
    family: str
    amount_requested: float | None
    submitted_at: str | None
    days_open: int
    # Against the partner's own avg_turnaround_days for this scheme.
    sla_days: int | None
    sla_state: Literal["ON_TRACK", "DUE", "BREACHED", "UNKNOWN"]
    applicant: ApplicantSummary
    readiness: ReadinessOut
    # Why the router sent this application here, carried from the routing decision.
    verdict: str | None = None
    matched_because: list[dict[str, Any]] = Field(default_factory=list)
    engine_version: str | None = None


class QueueResponse(BaseModel):
    partner_id: str
    partner_name: str
    is_currently_accepting: bool
    total: int
    items: list[QueueItem]
    counts_by_status: dict[str, int]


class PartnerActionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    action: Literal["ACKNOWLEDGE", "REQUEST_DOCS", "APPRAISE", "SANCTION", "DISBURSE", "REJECT"]
    reason: str | None = Field(default=None, max_length=500)


class CapacityRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    scheme_code: str | None = Field(
        default=None,
        description="Limit the change to one scheme. Omit to apply to every authorisation.",
    )
    is_currently_accepting: bool | None = None
    max_ticket: float | None = Field(default=None, gt=0)


class CapacityRow(BaseModel):
    scheme_code: str
    scheme_name: str
    family: str
    is_currently_accepting: bool
    min_ticket: float | None
    max_ticket: float | None
    avg_turnaround_days: int | None
    active_load: int | None


class CapacityResponse(BaseModel):
    partner_id: str
    partner_name: str
    rows: list[CapacityRow]


# --- analytics -----------------------------------------------------------------------


class FunnelStage(BaseModel):
    stage: str
    label: str
    count: int
    # Share of the stage before it. None for the first stage.
    conversion_from_previous: float | None = None


class FunnelOut(BaseModel):
    stages: list[FunnelStage]
    largest_drop_off: str | None
    largest_drop_off_pct: float | None


class MisroutingOut(BaseModel):
    """The primary KPI: applications that would have gone to the wrong counter."""

    redirects_suggested: int
    partners_filtered_for_authorisation: int
    partners_filtered_for_distance: int
    partners_filtered_for_ticket_size: int
    routing_calls: int
    total_prevented: int


class UnderservedDistrict(BaseModel):
    district: str
    state: str
    demand: int
    partners_within_25km: int
    nearest_authorised_km: float | None
    families_unserved: list[str]


class CoverageOut(BaseModel):
    radius_km: float
    districts_with_demand: int
    underserved: list[UnderservedDistrict]
    # Rendered as a bubble map rather than a choropleth; see the note in the README.
    points: list[dict[str, Any]]


class BreakdownRow(BaseModel):
    key: str
    label: str
    count: int
    share: float


class TurnaroundRow(BaseModel):
    partner_type: str
    applications: int
    median_days: float | None
    p90_days: float | None


class AnalyticsOut(BaseModel):
    generated_at: str
    engine_version: str
    funnel: FunnelOut
    misrouting: MisroutingOut
    coverage: CoverageOut
    scheme_mix: list[BreakdownRow]
    language_mix: list[BreakdownRow]
    turnaround: list[TurnaroundRow]
    status_mix: list[BreakdownRow]

File: apps/web/src/lib/console.ts
/**
 * Console API client and session handling.
 *
 * Separate from `lib/api.ts` on purpose. That file serves the citizen route, which has
 * a hard JS budget and no authentication at all; this one is for signed-in staff on a
 * desk. Keeping them apart means a console dependency can never leak into a citizen
 * bundle.
 *
 * The token lives in `sessionStorage`, not `localStorage`: closing the tab ends the
 * session, which is the right default for a shared branch terminal. It is deliberately
 * not a cookie — there is no CSRF surface to defend if the browser never attaches
 * credentials on its own.
 */
import { apiBase } from "@/lib/apiBase";

// Resolved per call: it depends on `window.location`, absent at module load during SSR.
const BASE = () => apiBase();

const TOKEN_KEY = "setu.console.token";
const USER_KEY = "setu.console.user";

export interface ConsoleUser {
  id: string;
  email: string;
  display_name: string;
  role: "CITIZEN" | "PARTNER" | "ADMIN";
  partner_id: string | null;
  partner_name: string | null;
}

export type Result<T> =
  | { ok: true; data: T }
  | { ok: false; error: string; status?: number };

function store(): Storage | null {
  try {
    return typeof window === "undefined" ? null : window.sessionStorage;
  } catch {
    // Blocked site data throws on access, not on use.
    return null;
  }
}

export function getToken(): string | null {
  try {
    return store()?.getItem(TOKEN_KEY) ?? null;
  } catch {
    return null;
  }
}

export function getUser(): ConsoleUser | null {
  try {
    const raw = store()?.getItem(USER_KEY);
    return raw ? (JSON.parse(raw) as ConsoleUser) : null;
  } catch {
    return null;
  }
}

export function signOut(): void {
  try {
    store()?.removeItem(TOKEN_KEY);
    store()?.removeItem(USER_KEY);
  } catch {
    /* nothing to clear */
  }
}

async function request<T>(path: string, init: RequestInit = {}): Promise<Result<T>> {
  const token = getToken();
  try {
    const response = await fetch(`${BASE()}${path}`, {
      ...init,
      headers: {
        ...(init.body ? { "Content-Type": "application/json" } : {}),
        ...(token ? { Authorization: `Bearer ${token}` } : {}),
        ...(init.headers ?? {}),
      },
    });
    if (!response.ok) {
      let detail = `Request failed (${response.status})`;
      try {
        const body = (await response.json()) as { detail?: string };
        if (body?.detail) detail = body.detail;
      } catch {
        /* keep the status-code message */
      }
      return { ok: false, error: detail, status: response.status };
    }
    return { ok: true, data: (await response.json()) as T };
  } catch {
    return { ok: false, error: "Could not reach the SamarthSetu API." };
  }
}

export async function signIn(email: string, password: string): Promise<Result<ConsoleUser>> {
  const result = await request<{ access_token: string; user: ConsoleUser }>("/auth/login", {
    method: "POST",
    body: JSON.stringify({ email, password }),
  });
  if (!result.ok) return result;
  try {
    store()?.setItem(TOKEN_KEY, result.data.access_token);
    store()?.setItem(USER_KEY, JSON.stringify(result.data.user));
  } catch {
    // A browser refusing storage still gets a working session for this page load.
  }
  return { ok: true, data: result.data.user };
}

// --- partner console -------------------------------------------------------------

export interface Readiness {
  score: number;
  uploaded: number;
  required: number;
  missing: string[];
  warnings: number;
  unmasked_documents: number;
}

export interface QueueItem {
  reference_no: string;
  status: string;
  scheme_code: string;
  scheme_name: string;
  family: string;
  amount_requested: number | null;
  submitted_at: string | null;
  days_open: number;
  sla_days: number | null;
  sla_state: "ON_TRACK" | "DUE" | "BREACHED" | "UNKNOWN";
  applicant: {
    display_name: string | null;
    gov_id_type: string | null;
    gov_id_last4: string | null;
    phone_last4: string | null;
    district: string | null;
    state: string | null;
    preferred_language: string;
  };
  readiness: Readiness;
  verdict: string | null;
  matched_because: { rule_id: string; message: string; severity: string }[];
  engine_version: string | null;
}

export interface QueueResponse {
  partner_id: string;
  partner_name: string;
  is_currently_accepting: boolean;
  total: number;
  items: QueueItem[];
  counts_by_status: Record<string, number>;
}

export interface CapacityRow {
  scheme_code: string;
  scheme_name: string;
  family: string;
  is_currently_accepting: boolean;
  min_ticket: number | null;
  max_ticket: number | null;
  avg_turnaround_days: number | null;
  active_load: number | null;
}

export interface CapacityResponse {
  partner_id: string;
  partner_name: string;
  rows: CapacityRow[];
}

export function getQueue(status?: string): Promise<Result<QueueResponse>> {
  const query = new URLSearchParams();
  if (status) query.set("status", status);
  return request<QueueResponse>(`/partner/queue${query.size ? `?${query}` : ""}`);
}

export function actOnApplication(
  reference: string,
  action: string,
  reason?: string,
): Promise<Result<QueueItem>> {
  return request<QueueItem>(`/partner/applications/${encodeURIComponent(reference)}/action`, {
    method: "POST",
    body: JSON.stringify({ action, ...(reason ? { reason } : {}) }),
  });
}

export function getCapacity(): Promise<Result<CapacityResponse>> {
  return request<CapacityResponse>("/partner/capacity");
}

export function setCapacity(body: {
  scheme_code?: string;
  is_currently_accepting?: boolean;
  max_ticket?: number;
}): Promise<Result<CapacityResponse>> {
  return request<CapacityResponse>("/partner/capacity", {
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

// --- admin dashboard -------------------------------------------------------------

export interface Analytics {
  generated_at: string;
  engine_version: string;
  funnel: {
    stages: {
      stage: string;
      label: string;
      count: number;
      conversion_from_previous: number | null;
    }[];
    largest_drop_off: string | null;
    largest_drop_off_pct: number | null;
  };
  misrouting: {
    redirects_suggested: number;
    partners_filtered_for_authorisation: number;
    partners_filtered_for_distance: number;
    partners_filtered_for_ticket_size: number;
    routing_calls: number;
    total_prevented: number;
  };
  coverage: {
    radius_km: number;
    districts_with_demand: number;
    underserved: {
      district: string;
      state: string;
      demand: number;
      partners_within_25km: number;
      nearest_authorised_km: number | null;
      families_unserved: string[];
    }[];
    points: Record<string, unknown>[];
  };
  scheme_mix: { key: string; label: string; count: number; share: number }[];
  language_mix: { key: string; label: string; count: number; share: number }[];
  status_mix: { key: string; label: string; count: number; share: number }[];
  turnaround: {
    partner_type: string;
    applications: number;
    median_days: number | null;
    p90_days: number | null;
  }[];
}

export function getAnalytics(): Promise<Result<Analytics>> {
  return request<Analytics>("/admin/analytics");
}

/** Fetches with the bearer token, then hands the browser a Blob — an <a href> cannot
 *  carry an Authorization header, so a plain link would 401. */
export async function downloadCsv(section: string): Promise<string | null> {
  const token = getToken();
  try {
    const response = await fetch(`${BASE()}/admin/export.csv?section=${section}`, {
      headers: token ? { Authorization: `Bearer ${token}` } : {},
    });
    if (!response.ok) return `Export failed (${response.status})`;
    const blob = await response.blob();
    const url = URL.createObjectURL(blob);
    const link = document.createElement("a");
    link.href = url;
    link.download = `setu-${section}.csv`;
    link.click();
    URL.revokeObjectURL(url);
    return null;
  } catch {
    return "Could not reach the SamarthSetu API.";
  }
}
